awx/main/tasks.py

```diff
@@ -149,6 +149,85 @@ def handle_setting_changes(self, setting_keys):
             break
 
 
+def _cancel_queues(app, worker_queues, worker_name):
+    removed_queues = []
+    for queue in worker_queues:
+        if queue['name'] in settings.AWX_CELERY_QUEUES_STATIC or \
+                {'alias': queue['name']} in settings.AWX_CELERY_QUEUES_STATIC:
+            continue
+
+        removed_queues.append(queue['name'])
+        app.control.cancel_consumer(queue['name'], reply=True, destination=[worker_name])
+    return removed_queues
+
+
+def _add_queues(app, instance, worker_name):
+    # Add worker to group-named queues
+    queues = []
+    for ig in instance.rampart_groups.all():
+        app.control.add_consumer(ig.name, reply=True, destination=[worker_name])
+        queues.append(ig.name)
+
+    # Add worker to worker-named queue
+    app.control.add_consumer(instance.hostname, reply=True, destination=[worker_name])
+    queues.append(instance.hostname)
+
+    return queues
+
+
+def _update_celery_routes(instance):
+    logger.debug("Updating celery routes for instance {}".format(instance.hostname))
+
+    tasks = [
+        'awx.main.tasks.cluster_node_heartbeat',
+        'awx.main.tasks.purge_old_stdout_files',
+    ]
+
+    # Instance is, effectively, a controller node
+    if instance.rampart_groups.filter(controller__isnull=False).exists():
```

Comment:
I think this is the test to determine if an instance is isolated, not if the instance is a controller.

To figure out if it's a controller, I think you need something different, using the reverse relationship. Like:

`Instance.objects.filter(rampart_groups__controller__instances=instance).exists()`

(no guarantees about the accuracy of this queryset)
You could also use the reverse `controlled_groups` relationship, but I can't put together how to use that in my head right now.